src/ir_passes/assume.rs

```diff
@@ -0,0 +1,133 @@
+use crate::{
+    ast,
+    ir::{self, Ctx, ExprIdx, PropIdx},
+    ir_visitor::{Action, Visitor},
+};
+
+/// Generates default assumptions to the Filament program for assumptions using custom functions
+#[derive(Default)]
+pub struct Assume;
+
+impl ir::Component {
+    /// Adds the assumptions associated with a proposition of the form `#l = f(#r)` to the component.
+    fn add_assumptions(
+        &mut self,
+        f: ast::UnFn,
+        lhs: ExprIdx,
+        rhs: ExprIdx,
+    ) -> Vec<PropIdx> {
+        // Define constant expressions used
+        let zero = self.add(ir::Expr::Concrete(0));
+        let one = self.add(ir::Expr::Concrete(1));
+        let two = self.add(ir::Expr::Concrete(2));
+
+        match f {
+            ast::UnFn::Pow2 => vec![
+                // #l * 2 = pow2(#r + 1)
+                lhs.mul(two, self)
+                    .equal(rhs.add(one, self).pow2(self), self),
+                // #r >= 1 => #l = pow2(#r - 1)*2
+                rhs.gte(one, self).implies(
+                    lhs.equal(
+                        rhs.sub(one, self).pow2(self).mul(two, self),
+                        self,
+                    ),
+                    self,
+                ),
+                // #l = 1 => #r = 0
+                lhs.equal(one, self).implies(rhs.equal(zero, self), self),
+                // #r = 0 => #l = 1
+                rhs.equal(zero, self).implies(lhs.equal(one, self), self),
```

Comment:
Ha! Isn't this syntax much nicer. If only we had [Scala's context parameters](https://docs.scala-lang.org/scala3/book/ca-context-parameters.html); we could make the `self` argument hidden too and it'd be perfect.